### ISSUE TITLE
Exposing analyzer.

### ISSUE THREAD

**Issue by yahu1031** [Dart-Code/Dart-Code#5008]
**Is your feature request related to a problem? Please describe.**
Not exactly a problem, but extensions depending on dart LSP will make this feature request more profitable. Exposing required/specific data would be more helpful.

**Describe the solution you'd like**
Exposing analyzer.

**Additional context**
In my case, we are building the [fluttergpt](https://marketplace.visualstudio.com/items?itemName=WelltestedAI.fluttergpt) extension, which will utilize the dart LSP. [This](https://github.com/Dart-Code/Dart-Code/commit/d8f943d8bcc10834262155a057ae412fb29fab52) had become a blocker for us. Ref: https://github.com/Welltested-AI/dash-ai/issues/224

**Comment by DanTup** [comment]
@yahu1031 can you add details of exactly what functionality you're using (code links would be helpful too) so I can understand exactly what would need to be provided if "raw" access to the analyzer was taken away?

A brief description of _why_ you need it (eg. what functionality it powers) would be useful (not just which specific LSP requests you're calling).

If there are things you're not _currently_ using, but are considering or might be helpful, please include details too.

Thanks!

**Comment by samyakkkk** [comment]
@DanTup yes, we are building an AI extension that provides extra abilities to Flutter developers on top of the current existing Dart extension. 

We found it really promising that the internal APIs like the analyzer (that we currently use to get the Flutter outline), and other like devtools, pubGlobal (that we plan to use) are exposed making it easy for us to call these methods from our extension. Loosing access to those will require us to duplicate the same code on our end. 

On a separate issue, we had a brief discussion with Jacob and others in the Flutter team. We wanted to figure out a way that we could get direct access to the Dart Server spun up by the VSCODE extension because a duplicate instance on our end is not feasible because of the memory issue. We run custom kind of analysis on the files for various purposes. 

I could start a separate thread/issue for the dart server access extension issue, but in general the exposed internal APIs could come really handy to us in future and would love if they are kept exposed. (maybe in a safer way). 

Thanks.

**Comment by DanTup** [comment]
@samyakkkk the problem with the internal APIs is that they are not designed for consumption and have no versioning. They frequently change (particularly when things are refactored) and will break your extension (like this recent change).. this is not a problem for the tests they are intended for because those are versioned in the same repository.

I don't want to break any of your functionality, but it's not feasible to officially support all of the current private APIs (plus anything that is added in future to simplify for testing), the surface area is just too large and volatile (it's just internal classes exposed). If other extensions need access to things, we need to expose them as a supported API that we can commit to keeping stable.

For the LSP server specifically, the plan we've discussed before is to block access to state-modifying requests that would break the Dart extension (like `textDocument/*`). My expectation is that you are not using these APIs so I think your use of the LSP server will fit within the APIs that we'd like to make available. However, I can't confirm that without more details about what you're using.

If you're also using other (non-LSP) APIs (you mentioned Pub?), we also need to know which to see what would be required to provide a stable API over them.

Thanks!

**Comment by samyakkkk** [comment]
Understood the point you are making @DanTup . It seems promising to add the required support to the newly exposed VScode APIs. 

I also read the notifications documentation but that may not be the solution for us, since our requirements are more of actively seeking information and triggering actions, than listening passively for events. However, I agree with the intent here and can open new issue(s) documenting with our requirements, for considering if they can be officially support. 

Thanks again for keeping the currently exposed APIs intact. I apologise for the trouble on this. We didn't consider this case and became too much dependent on the internal APIs :)

**Comment by tomgilder** [comment]
Hello, for [Wings](https://flutterwings.dev) I'm using two private APIs, which I think should be very viable to support publicly.

They're just the SDK roots, in order to know where the currently-in-use Flutter and Dart installs are: `workspaceContext.sdks.flutter` and `workspaceContext.sdks.dart`.

**Comment by DanTup** [comment]
@tomgilder thanks! Those should be easy to handle - although some things worth noting:

- The user can change their SDK while the extension is running (this triggers a reload of the Dart extension internally) so probably we need to provide some kind of event/stream for this so you can re-read the SDK if required
- When a Flutter project is open, the Dart/Flutter SDKs will be consistent (that is, the Dart SDK will be inside the Flutter SDK). However when a non-Flutter project is open, they may be different (because we're tracking the Flutter SDK only for if the user tries to execute Flutter commands like "Flutter: New Project" but may have used a preferred Dart SDK that isn't part of a Flutter SDK)

**Comment by DanTup** [comment]
@yahu1031 @samyakkkk @tomgilder sorry for the delay with this. I'm currently looking at exposing a better API that will be supported (with the goal of ultimately hiding all the internals).

So far the API includes the ability to get the DTD URI and the SDK paths/versions and events for when they change (see interfaces below).

Since this issue is a little old, I wanted to check again what functionality you might be using to avoid implementing anything that isn't required. Some LSP functionality will be available over DTD (since we register many LSP methods there) and I don't want to also expose this here.

@samyakkkk you previous mentioned you're using `waitForOutline` which is not part of the new API, however I believe you can get this data through the [VS Code command](https://code.visualstudio.com/api/references/commands) `vscode.executeDocumentSymbolProvider`, so I also don't think we need to expose that.

Please let me know if you think anything besides the below is required to maintain existing functionality you have (at this point, I'm focusing on trying to hide the internals and not on providing new functionality).

```ts
import * as vs from "vscode";

export interface PublicDartExtensionApi {
	/**
	 * The current version of the public API.
	 */
	readonly version: number;

	/**
	 * The currently detected Dart and Flutter SDK paths and versions.
	 */
	readonly sdks: PublicSdks;

	/**
	 * An event that fires when the detected SDKs change. This may happen if the
	 * user changes their configuration, for example.
	 *
	 * An event does not necessarily mean the SDK paths have changed, only that they _might_ have.
	 */
	readonly onSdksChanged: vs.Event<PublicSdks>;

	/**
	 * The current Dart Tooling Daemon (DTD) URI, if available.
	 */
	readonly dtdUri: string | undefined;

	/**
	 * An event that fires when the DTD URI changes.
	 */
	readonly onDtdUriChanged: vs.Event<string | undefined>;
}

export interface PublicSdks {
	/**
	 * The absolute path to the Dart SDK folder, if detected.
	 */
	readonly dart?: string;

	/**
	 * The Dart SDK version string, as read from the Dart SDK's version file.
	 */
	readonly dartVersion?: string;

	/**
	 * The absolute path to the Flutter SDK folder, if detected.
	 */
	readonly flutter?: string;

	/**
	 * The Flutter SDK version string, as read from the Flutter SDK's version file.
	 */
	readonly flutterVersion?: string;
}
```

**Comment by samyakkkk** [comment]
Hi @DanTup , our project was shut down - so we are not currently using the APIs. Please feel free to exclude our earlier mentioned requirements.

Still very exciting to see this is being worked on!

**Comment by DanTup** [comment]
@samyakkkk ah, sorry to hear!

I'll proceed with the changes to add SDKs + DTD URI (PR at https://github.com/Dart-Code/Dart-Code/pull/5583) and consider this done for now (we can file new issues if there are requirements for additional things in future).

@tomgilder let me know if there's anything else you need, or if you need any pointers switching to this. You'll need to check `version >= 2` on the `exports` of this extension to know if the new fields are available (and I can publish a pre-release that includes these if you want to work on it prior to the next stable release which is end of month/start of next month).

**Comment by DanTup** [comment]
I've merged the change to add the SDKs and DTD URI. I've opened https://github.com/Dart-Code/Dart-Code/issues/5587 to track removal of the old APIs, which includes a list of known extensions and which APIs they are using (and links to issues I've filed in their repos).

If you know of any other extensions that are using these APIs, please ping a comment in that issue so we can try to ensure they have alternatives before they are removed.

Thanks!
